plugins/woocommerce/src/Internal/DataStores/Orders/LegacyDataHandler.php

```diff
@@ -195,6 +197,137 @@ private function is_order_newer_than_post( $order ): bool {
+	 */
+	public function get_diff_for_order( int $order_id ): array {
+		$diff = array();
+
+		$hpos_order = $this->get_order_from_datastore( $order_id, 'hpos' );
+		$cpt_order  = $this->get_order_from_datastore( $order_id, 'cpt' );
+
+		if ( $hpos_order->get_type() !== $cpt_order->get_type() ) {
+			$diff['type'] = array( $hpos_order->get_type(), $cpt_order->get_type() );
+		}
+
+		$hpos_meta = $this->order_meta_to_array( $hpos_order );
+		$cpt_meta  = $this->order_meta_to_array( $cpt_order );
+
+		// Consider only keys for which we actually have a corresponding HPOS column or are meta.
+		$all_keys = array_unique(
+			array_diff(
+				array_merge(
+					$this->get_order_base_props(),
+					array_keys( $hpos_meta ),
+					array_keys( $cpt_meta )
+				),
+				$this->data_synchronizer->get_ignored_order_props()
+			)
+		);
+
+		foreach ( $all_keys as $key ) {
+			$val1 = in_array( $key, $this->get_order_base_props(), true ) ? $hpos_order->{"get_$key"}() : ( $hpos_meta[ $key ] ?? null );
+			$val2 = in_array( $key, $this->get_order_base_props(), true ) ? $cpt_order->{"get_$key"}() : ( $cpt_meta[ $key ] ?? null );
+
+			// Workaround for https://github.com/woocommerce/woocommerce/issues/43126.
+			if ( ! $val2 && in_array( $key, array( '_billing_address_index', '_shipping_address_index' ), true ) ) {
+				$val2 = get_post_meta( $order_id, $key, true );
+			}
+
+			if ( $val1 != $val2 ) { // phpcs:ignore WordPress.PHP.StrictComparisons.LooseComparison
+				$diff[ $key ] = array( $val1, $val2 );
+			}
+		}
+
+		return $diff;
+	}
+
+	/**
+	 * Returns an order object as seen by either the HPOS or CPT datastores.
+	 *
+	 * @since 8.6.0
+	 *
+	 * @param int    $order_id      Order ID.
+	 * @param string $data_store_id Datastore to use. Should be either 'hpos' or 'cpt'. Defaults to 'hpos'.
+	 * @return \WC_Order Order instance.
+	 */
+	public function get_order_from_datastore( int $order_id, string $data_store_id = 'hpos' ) {
+		$data_store = ( 'hpos' === $data_store_id ) ? $this->data_store : $this->data_store->get_cpt_data_store_instance();
+
+		wp_cache_delete( \WC_Order::generate_meta_cache_key( $order_id, 'orders' ), 'orders' );
 
+		// Prime caches if we can.
+		if ( method_exists( $data_store, 'prime_caches_for_orders' ) ) {
+			$data_store->prime_caches_for_orders( array( $order_id ), array() );
+		}
+
+		$classname = wc_get_order_type( $data_store->get_order_type( $order_id ) )['class_name'];
+		$order     = new $classname();
+		$order->set_id( $order_id );
+
+		// Switch datastore if necessary.
+		$update_data_store_func = function ( $data_store ) {
+			// Each order object contains a reference to its data store, but this reference is itself
+			// held inside of an instance of WC_Data_Store, so we create that first.
+			$data_store_wrapper = \WC_Data_Store::load( 'order' );
+
+			// Bind $data_store to our WC_Data_Store.
+			( function ( $data_store ) {
```

Comment:
Interesting pattern! Could you explain why this was the best (only?) option here? Just for my personal understanding.

Reply:
Given `WC_Order` is a `WC_Data` child class, the datastore in objects of this type [has access `protected`](https://github.com/woocommerce/woocommerce/blob/ea7e97caea9695c898a0cf2e6efd0176440cb8b9/plugins/woocommerce/includes/abstracts/abstract-wc-data.php#L91) so we can't directly change it from the outside.
To make things worse, that `data_store` variable is not directly an instance of the HPOS (or CPT) datastore, but of `WC_Data_Store`, which [stores the details in a `private` variables](https://github.com/woocommerce/woocommerce/blob/75721ed9de09445178d97fcbb5e54d88a34feccb/plugins/woocommerce/includes/class-wc-data-store.php#L17-L63).

`WC_Data_Store` is meant to be a singleton so what we could do here is add public methods to `WC_Data` to change its backing datastore, but that's not without its own set of problems (what happens when reading or saving data if you change the backing datastore in an object mid execution?) and I guess we don't want to increase API surface at this moment.

Now, even though datastores have [a `read()` method that receives a "blank" object](https://github.com/woocommerce/woocommerce/blob/9ce04dc807805d1c1638acc1a7e3fe08adfe69b1/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php#L1116), this method's name is a bit misleading because this will not set the backing datastore on the blank object nor read all properties.
Metadata [is read via `$this->data_store->read_meta()`](https://github.com/woocommerce/woocommerce/blob/ea7e97caea9695c898a0cf2e6efd0176440cb8b9/plugins/woocommerce/includes/abstracts/abstract-wc-data.php#L646) which would use the "default" datastore unless we change it "somehow".

This ugly hack above is that "somehow" 😓. I believe the whole CRUD layer and datastore/object interactions in particular would benefit from a revision, but 🤷.

(Sorry if this is even more confusing)